### PR TITLE
[WIP] Hi Haddock

### PR DESCRIPTION
**Pull request by sjakobi**
([GSoC project description](https://summerofcode.withgoogle.com/projects/#4975710121230336))

## What already works

For many packages we can already generate reasonably-looking haddocks. For example see these [haddocks for lens](https://github.com/haskell/haddock/files/2286675/lens-4.17-docs.tar.gz).

## TODOs

* [x] ~~Get `html-test` to run. Currently we can't handle the `Bug294` test-case. This requires changes in `extractDecl`. ([Details](https://github.com/sjakobi/haddock/issues/25))~~ Fixed by @harpocrates in https://github.com/haskell/haddock/pull/906/commits/29edbb9d9e085afcf947512bb149de5492623aaa.

* [ ] Scrutinize output changes and update the tests. In particular we currently generate kind signatures for all higher-kinded type variables. ([Details](https://github.com/sjakobi/haddock/issues/5))

* [x] Fix `--hyperlinked-source` by integrating Zubin Duggal's GSoC project.

* [ ] Warnings for ambiguous identfiers are too noisy. ([Details](https://github.com/sjakobi/haddock/issues/21))

A few more smaller issues are marked `TODO` in the source code.

---

The corresponding changes in GHC are [on Phabricator](https://phabricator.haskell.org/D5067).

### CODE REVIEW

**Comment by harpocrates** [comment]
I've started looking at the test outputs. One of the bigger issues is that  the output changes depending on whether there is already an .hi file:

```
haddock$ rm html-test/src/*hi
haddock$ cabal new-test html-test 2> /dev/null | grep FAIL | wc -l
21
haddock$ cabal new-test html-test 2> /dev/null | grep FAIL | wc -l
31
```

I've found at least two causes for this:

  * source locations of instances are dropped if there is already a .hi file (these locations are supposed to be found on the spans of the third element in the `DocInstance` tuple). In the test cases, this manifests as a missing `#`. 
  * type variables can be inconsistently named (without an .hi file, type variable names can shadow outer variables but with an .hi file, these get given fresh names).

**Comment by harpocrates** [comment]
The logic for figuring out which modules to load for attaching instances also needs to be ported over. Some issues about instances missing are back. 😢

An easy to spot example of this: the `Array` type from `Data.Array` is missing its `Foldable`, `Traversable`, and `Read` instances.

**Comment by harpocrates** [comment]
Note for when this gets merged to `ghc-head`: https://github.com/haskell/haddock/pull/906/commits/637e14b23b193e8a2d7084ddf6a9cc274a60942a should have fixed https://ghc.haskell.org/trac/ghc/ticket/11955.

**Comment by harpocrates** [comment]
@sjakobi Are you still actively working on this? If not, do you mind if I rebase this + the Phab diff? I would also like to go ahead and fix some of the outstanding issues on the Phab diff, hopefully getting it merge ready...

**Comment by harpocrates** [comment]
Looks like my email reply from last night fell victim to GitHub's downtime:

---

I've updated this branch as well as your GHC changes. I don't think I can update the Phabricator diff without comandeering it (I'm not even sure I can, now that I think about it 😄), so I've left the rebased changes here: https://github.com/harpocrates/ghc/tree/arcpatch-D5067-rebased. Putting those up on Phabricator should be a matter of

    ghc$ git checkout arcpatch-D5067-rebased
    ghc$ arc diff cb98646cd67c85c2b2b33504a336d841f1ea70e5^ --update D5067

Cheers!

**Comment by sjakobi** [comment]
@harpocrates I've updated the Phab diff according to your instructions. Thanks a lot for the rebase!

I actually wouldn't mind if you'd comandeer the diff BTW.

**Comment by sjakobi** [comment]
@harpocrates Could you give https://github.com/haskell/haddock/pull/906/commits/cad2ed97005149f52012550a75e0f8cde8e5e831 a quick review?

The main thing I'm worried about is that in `hsDocRenamer`, I might recreate the `env` for each identifier I'm looking up. I'm just not sure what kind of code GHC generates there… :/

https://github.com/haskell/haddock/blob/cad2ed97005149f52012550a75e0f8cde8e5e831/haddock-api/src/Haddock/Interface/LexParseRn.hs#L193-L196

File: haddock-api/src/Haddock/Convert.hs [hunk 1]
```diff
@@ -423,22 +496,36 @@ data SynifyTypeState
...
+defaultType HideRuntimeRep = defaultRuntimeRepVars
+
+-- | Convert a core type into an 'HsType'.
+synifyType
+  :: SynifyTypeState  -- ^ what to do with a 'forall'
+  -> [TyVar]          -- ^ free variables in the type to convert
+  -> Type             -- ^ the type to convert
+  -> LHsType GhcRn
+synifyType _ _ (TyVarTy tv)
+  | mkTyVarOcc "_" == occName tv = noLoc $ HsWildCardTy $ AnonWildCard n
```
Comment (sjakobi) [inline haddock-api/src/Haddock/Convert.hs]: @mynguyenbmc @harpocrates I'm not sure what to do about this line when merging the visible [kind application patch](https://github.com/haskell/haddock/commit/21e4f3fa6f73a9b25f3deed80da0e56024238ea5) into my branch. Can you help me out?

File: haddock-api/src/Haddock/Convert.hs [hunk 2]
```diff
@@ -426,22 +499,36 @@ data SynifyTypeState
...
+defaultType HideRuntimeRep = defaultRuntimeRepVars
+
+-- | Convert a core type into an 'HsType'.
+synifyType
+  :: SynifyTypeState  -- ^ what to do with a 'forall'
+  -> [TyVar]          -- ^ free variables in the type to convert
+  -> Type             -- ^ the type to convert
+  -> LHsType GhcRn
+synifyType _ _ (TyVarTy tv)
+  | mkTyVarOcc "_" == occName tv = error "synifyType: FIXME" -- noLoc $ HsWildCardTy $ AnonWildCard n
```
Comment (sjakobi) [inline haddock-api/src/Haddock/Convert.hs]: @harpocrates @mynguyenbmc This is my "workaround" for now. I haven't actually seen this error pop up though. If we keep this special case we should probably add a testcase.

Comment (mynguyenbmc) [reply haddock-api/src/Haddock/Convert.hs]: @sjakobi I would think noLoc $ HsWildCardTy noExt would work, have you tried it?

Comment (sjakobi) [reply haddock-api/src/Haddock/Convert.hs]: Cheers @mynguyenbmc! That seems to work! :+1:

**Comment by sjakobi** [comment]
It's good that you're asking, @bgamari. [!1648](https://gitlab.haskell.org/ghc/ghc/merge_requests/1648#note_220038) reminded me of some pretty cool work by @harpocrates among the other Hi Haddock stuff that resulted in proper haddocks for `[]` and `(->)`.

I had got stuck on #921 last winter, and had little time for this project. In March I ended up asking @harpocrates to take over the remaining work.

Back then I was under the impression that there were still quite a few tricky (tricky for me) issues to resolve. AFAIK the latest TODO list is [this one](https://gitlab.haskell.org/ghc/ghc/merge_requests/608/#note_189219). What made everything more difficult was the need to sync (and rebase) things between haddock and ghc.

I'm wondering whether there could be a way to merge parts of the work incrementally, reducing the amount of changes outside of master that have to be kept in sync.

@harpocrates what's your perspective on all this these days?

**Comment by harpocrates** [comment]
Sorry, I'm definitely at fault for dragging my feet on this. I thought I'd negotiated 2 weeks of paid time for this with my employer last April, but that ended up not panning out. Lately I have a bit more time - perhaps even as early as this month...

  * #921 isn't of immediate concern. I think initially Hi Haddock should unconditionally re-generate `.hi` files with docstrings (the same way it currently unconditionally re-generates `.hie` files for hyperlinking). When we get to the optimization where Haddock potentially re-uses `.hi`/`hie` files, we'll have to deal with #921. Besides, that line of progress will require some work on Cabal too.

  * I've already cherry-picked a bunch of the improvements that initially started in the `hi-haddock` branch of Haddock in an effort to trim this PR down (#931, #938, #1003, #1005, #1022, #1020, #1045, and probably some more). Post-rebase, I expect the Haddock diff to be relatively small.

  * The Haddock-side is done, except for rebasing. The remaining work is really the TODO list (thank @sjakobi for the link): https://gitlab.haskell.org/ghc/ghc/merge_requests/608/#note_189219.

**Comment by Kleidukos** [comment]
What is our current status? Is there a way to better split this PR?

**Comment by Kleidukos** [comment]
@harpocrates @sjakobi

**Comment by Kleidukos** [comment]
Hi!  As of today, patches for the `ghc-head` branch must be done on the Haddock mirror on GitLab, at https://gitlab.haskell.org/ghc/haddock.

**Pull request closed without merging**